### PR TITLE
Use address type (IP vs DNS) configured by ListenerClass

### LINKED ISSUE #139: Allow listenerclasses to specify whether to prefer IP or Hostname addresses

### PR DESCRIPTION
**Pull request by nightkr**
# Description

Fixes #139, requires https://github.com/stackabletech/operator-rs/pull/885

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [x] CRD changes approved
- [x] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] Code contains useful logging statements
- [x] (Integration-)Test cases added
- [x] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [x] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,16 +4,22 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased]
 
+### Added
+
+- Listener.status.addresses now allows users to configure whether they prefer IP addresses or DNS hostnames ([#233]).
```
Comment (sbernauer) [inline CHANGELOG.md:9]:
```suggestion
- `Listener.spec.preferredAddressType` now allows users to configure whether they prefer IP addresses or DNS hostnames ([#233]).
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,16 +4,22 @@ All notable changes to this project will be documented in this file.
...
+
 ### Changed
 
 - Listener.status.addresses for NodePort listeners now includes replicas that are currently unavailable ([#231]).
+- Listener.status.addresses now defaults to DNS hostnames for all service types (previously NodePort and ClusterIP would prefer IP addresses, [#233]).
```
Comment (sbernauer) [inline CHANGELOG.md:14]:
```suggestion
- `Listener.status.addresses` now defaults to DNS hostnames for all service types (previously NodePort and ClusterIP would prefer IP addresses [#233]).
```

File: Cargo.toml [hunk 3]
```diff
@@ -31,5 +31,7 @@ tonic-build = "0.12"
 tonic-reflection = "0.12"
 tracing = "0.1.40"
 
-# [patch."https://github.com/stackabletech/operator-rs.git"]
+[patch."https://github.com/stackabletech/operator-rs.git"]
+# stackable-operator = { path = "../operator-rs/crates/stackable-operator" }
 # stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "feature/listenerclass-preferred-address-type" }
```
Comment (sbernauer) [inline Cargo.toml:37]: Marker to switch to release

File: rust/operator-binary/src/listener_controller.rs [hunk 4]
```diff
@@ -269,6 +272,7 @@ pub async fn reconcile(listener: Arc<Listener>, ctx: Arc<Ctx>) -> Result<control
         })?;
 
     let nodes: Vec<Node>;
+    let kubernetes_service_fqdn: String;
```
Comment (sbernauer) [inline rust/operator-binary/src/listener_controller.rs:275]: Getting C vibes ^^

File: rust/operator-binary/src/listener_controller.rs [hunk 5]
```diff
@@ -321,13 +327,19 @@ pub async fn reconcile(listener: Arc<Listener>, ctx: Arc<Ctx>) -> Result<control
...
+            addresses = match listener_class.spec.preferred_address_type {
+                AddressType::Ip => svc
+                    .spec
+                    .iter()
+                    .flat_map(|s| &s.cluster_ips)
+                    .flatten()
+                    .map(|addr| (&**addr, AddressType::Ip))
+                    .collect::<Vec<_>>(),
+                AddressType::Hostname => {
+                    kubernetes_service_fqdn = format!("{svc_name}.{ns}.svc.cluster.local");
```
Comment (sbernauer) [inline rust/operator-binary/src/listener_controller.rs:339]: FYI @maltesander this is a new place to add to your list (in case you have one)

**Review by sbernauer — CHANGES_REQUESTED** [review]
LGTM in general, did not test it yet

File: CHANGELOG.md (hunk 1 above)
Comment (nightkr) [reply CHANGELOG.md:9]: The configuration field is in `ListenerClass`. It should probably be something closer to "`Listener.status.addresses` can now be configured to..."

File: rust/operator-binary/src/listener_controller.rs [hunk 6]
```diff
@@ -321,13 +327,19 @@ pub async fn reconcile(listener: Arc<Listener>, ctx: Arc<Ctx>) -> Result<control
...
-            addresses = svc
-                .spec
-                .iter()
-                .flat_map(|s| &s.cluster_ips)
-                .flatten()
-                .map(|addr| (&**addr, AddressType::Ip))
-                .collect::<Vec<_>>();
+            addresses = match listener_class.spec.preferred_address_type {
+                AddressType::Ip => svc
+                    .spec
```
[the remark this answers is not in the document]
Comment (nightkr) [reply rust/operator-binary/src/listener_controller.rs:332]: Yup.. Outside of our control, I'm afraid.

File: rust/operator-binary/src/utils.rs [hunk 7]
```diff
@@ -101,26 +101,42 @@ pub fn error_full_message(err: &dyn std::error::Error) -> String {
...
+        match preferred_address_type {
+            AddressType::Ip => ip.or(hostname),
+            AddressType::Hostname => hostname.or(ip),
+        }
+    }
+}
+
 /// Try to guess the primary address of a Node, which it is expected that external clients should be able to reach it on
-pub fn node_primary_address(node: &Node) -> Option<(&str, AddressType)> {
+pub fn node_primary_address(node: &Node) -> AddressCandidates {
```
[the remark this answers is not in the document]
Comment (nightkr) [reply rust/operator-binary/src/utils.rs:122]: I don't know, those are exactly the kind of nothing-tests that tend to bring far less value than what they cost (in terms of readability especially, but also maintenance).

File: rust/operator-binary/src/listener_controller.rs (hunk 4 above)
Comment (nightkr) [reply rust/operator-binary/src/listener_controller.rs:275]: Yuup, they've gotta be stored somewhere though.

File: CHANGELOG.md (hunk 1 above)
Comment (nightkr) [reply CHANGELOG.md]: 9c2ce25586bacc779e0dfedd95d674b704ddf023

Comment (sbernauer) [reply CHANGELOG.md]: works for me

**Review by sbernauer — DISMISSED** [review]
LGTM, we only need the op-rs release

**Review by sbernauer — APPROVED** [review]
Alright, let's use main for now

**Comment by sbernauer** [comment]
I would like to start a weekly test-run for all products after the merge, so that we see if it causes any problems.

**Comment by sbernauer** [comment]
@nightkr good news: I ran 5 tests ,only https://testing.stackable.tech/job/hdfs-operator-it-weekly/17/ failed - which seems like flakiness :)
`Get "[https://cluster-dns-n9acebml.hcp.eastus.azmk8s.io:443/apis/batch/v1/namespaces/kuttl-test-wealthy-liger/jobs/access-hdfs](https://cluster-dns-n9acebml.hcp.eastus.azmk8s.io/apis/batch/v1/namespaces/kuttl-test-wealthy-liger/jobs/access-hdfs)": dial tcp: lookup cluster-dns-n9acebml.hcp.eastus.azmk8s.io on 8.8.8.8:53: no such host`

**Comment by nightkr** [comment]
:/

**Pull request merged**